File: my-solutions/0138-copy-list-with-random-pointer/solution.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        # without hashmap
        if not head:
            return None

        dummyAdder = head
        while(dummyAdder.next):
            temp = dummyAdder.next
            dummyAdder.next = Node(dummyAdder.val)
            dummyAdder.next.next = temp
            dummyAdder = dummyAdder.next.next

        dummyAdder.next = Node(dummyAdder.val)

        # temp = head
        # while(temp):
        #     print(temp, temp.val)
        #     temp = temp.next
        # return head
        
        randomAdder = head
        while(randomAdder and randomAdder.next):
            temp = randomAdder.next.next
            if randomAdder.random:
                randomAdder.next.random = randomAdder.random.next
            randomAdder = temp

        splitter = head
        res = head.next
        while(splitter.next.next):
            temp = splitter.next.next
            splitter1 = splitter
            splitter2 = splitter.next
            splitter1.next = splitter1.next.next
            splitter2.next = splitter2.next.next
            splitter = temp
        
        return res

File: my-solutions/0138-copy-list-with-random-pointer/solution.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        unordered_map<Node*, Node*> nodeMap{};
        Node* prev = nullptr;
        Node* res = head;
        while(head) {
            if(!nodeMap.count(head)) {
                nodeMap[head] = new Node(head->val);
            }
            if(head->random && !nodeMap.count(head->random)) {
                nodeMap[head->random] = new Node(head->random->val);
            }
            nodeMap[head]->random = nodeMap[head->random];
            if(prev) nodeMap[prev]->next = nodeMap[head];
            prev = head;
            head = head->next;
        }
        return nodeMap[res];
    }
};
